### ISSUE TITLE
Implement new endpoint that automatically works out query or execute

### ISSUE THREAD

**Issue by otoolep** [rqlite/rqlite#263]
Having to explicitly choose the API endpoint, depending on the command type, is awkward. A new endpoint, that contained a built-in SQL parser, would be best. It would route the statements to the correct endpoints (or internal calls) as needed.

**Comment by otoolep** [comment]
If a full SQLite parser was built in, rqlite could also support a streaming TCP connection for getting statements into the system. This might also improve throughput.

Anyone know of SQLite command parsing code? Perhaps I could pull in the source code of the SQLite system itself, unless it's already in the library I am using.

**Comment by otoolep** [comment]
"cznic/ql have own lexer for SQL and it syntax very similar."

**Comment by kenshaw** [comment]
I'm not going to write a full parser. Instead, I'm just going to fake it. As far as I can tell, there's only about 8 or 9 different "beginning" words that should be one versus the other. At the end of the day with rqlite, you might want to access the actual underlying sqlite libraries directly. Sqlite has a full blown parser built in, so I would imagine it'd be relatively straightforward to start from there?

Originally, when I was planning usql, I was thinking I would bind to the postgres parser. But that's icky.

**Comment by otoolep** [comment]
@kenshaw -- I think you meant to add this comment to the `usql` repo, not sure. No worries.

>>As far as I can tell, there's only about 8 or 9 different "beginning" words that should be one versus the other.
I don't know for sure. You might be correct. It seems statements that begin with `WITH` may only have `SELECT` statements following. It looks like it, at least for SQLite: https://www.sqlite.org/lang_with.html

**Comment by otoolep** [comment]
Actually, no that's not true -- it seems `INSERT` can use a `WITH` as well. So the idea that checking "starting" words will be enough doesn't seem right.

https://www.sqlite.org/syntax/insert-stmt.html

**Comment by otoolep** [comment]
In other words, look at these two statements:
```
With helper_table As (
Select * From dummy2
)
Insert Into dummy1 Values (Select t.a From helper_table t Where t.a = 'X' );
```
```
WITH Employee_CTE (EmployeeNumber, Title)
AS
(SELECT NationalIDNumber,
        JobTitle
 FROM   HumanResources.Employee)
SELECT EmployeeNumber,
       Title
FROM   Employee_CTE
```
Each should go to a different handler, even though both start with `WITH`.

**Comment by kenshaw** [comment]
Yes, I'm deeply embarrassed. I had too many tabs open simultaneously. Apologies!

**Comment by kenshaw** [comment]
@otoolep Just want to point out that I wrote some notes on the [`usql` issue](https://github.com/knq/usql/issues/3). I think for `rqlite`, you might be able to get away with the same trick I am using in `xo` to inspect queries:

```sql
CREATE TEMPORARY VIEW __random_id AS <query>
```

**Comment by otoolep** [comment]
A Pratt Parser might work well for this.

**Comment by otoolep** [comment]
Basically rqlite needs a Go module that can robustly parse a single SQL command and tell whether it will modify the database or not. Currently the CLI just checks with the command starts with `SELECT`, an approach which covers most cases, but not all. 

This Go module could then be integrated into the rqlite software itself.

**Comment by otoolep** [comment]
This looks useful: https://github.com/xwb1989/sqlparser

**Comment by gunnihinn** [comment]
Would it be acceptable to use [`stmt_readonly`](https://www.sqlite.org/c3ref/stmt_readonly.html) to detect this?

**Comment by otoolep** [comment]
That's interesting, I didn't know about that statement. It may do what I need -- thanks for the pointer.

**Comment by chermehdi** [comment]
How about generating the SQL parser using something like antlr, there is an existing grammar for the sqlite specific syntax: https://github.com/antlr/grammars-v4/blob/master/sql/sqlite/SQLiteParser.g4

I guess this wins 2 things: 

* You can determine the type of query.
* You can pre-validate the query before propagating to other raft nodes which could reduce the amount of garbage that you store in the Raft Log.

**Comment by otoolep** [comment]
Yes, adding an actual SQLite parser to the system was something I was considering. It would help rqlite as you say.

**Comment by otoolep** [comment]
https://developers.cloudflare.com/d1/platform/client-api/

**Comment by otoolep** [comment]
I would like some comment on this proposal. It will allow clients to send read or write SQL -- or a mix -- to a single endpoint and rqlite will figure out what to do.

My proposal for this endpoint is as follows:

**Path:** `/db/request`
**Methods:** `POST`

The request body will be a JSON array, each element is SQL statement. Straight SQL, parameterized SQL, and Named Parameterized SQL will all be supported. The request body can be a mix of both read and write statements. For example:

```json
[
	"INSERT INTO foo(name) VALUES('fiona')",
	["INSERT INTO foo(name) VALUES(?)", "declan"],
	"SELECT * FROM foo WHERE name='fiona'",
	["SELECT * FROM foo WHERE name=?", "dana"]
]
```

For each object in the top-level array rqlite will determine if it's a read-only statement, as determined by the SQLite function [`sqlite3_stmt_readonly`](https://www.sqlite.org/c3ref/stmt_readonly.html). If a read consistency level is passed as a URL query parameter, it will be applied to any read-only statements i.e. queries.

The response object will be a of the following structure, which is an object per SQL statement in the results array:

```json
{
    "results": [
        {
            "last_insert_id": 1,
            "rows_affected": 1,
            "time": 0.00759015
        },
        {
            "last_insert_id": 2,
            "rows_affected": 1,
            "time": 0.00669015
        },
        {
            "columns": [
                "id",
                "name",
                "age"
            ],
            "types": [
                "integer",
                "text",
                "integer"
            ],
            "values": [
                [
                    1,
                    "fiona",
                    20
                ]
            ],
            "time": 0.0150043
        },
        {
            "columns": [
                "id",
                "name",
                "age"
            ],
            "types": [
                "integer",
                "text",
                "integer"
            ],
            "values": [
                [
                    1,
                    "fiona",
                    20
                ]
            ],
            "time": 0.0150043
        }
    ],
    "time": 0.869015
}
```
Clients will need to inspect each object in the results array to determine whether it's a read response or write response.

**Comment by jmordica** [comment]
Thanks. 

1. Will clients be able to specify `associative` in the request?
2. Will clients be able to specify read consistency?

**Comment by otoolep** [comment]
To quote from above: _If a read consistency level is passed as a URL query parameter, it will be applied to any read-only statements i.e. queries_.

As for associative, that should be fine, again only applied to the query requests to start.

**Comment by jmordica** [comment]
For sure. Both comments are understood.

**Comment by otoolep** [comment]
Couple of other specification points:

- This API is only for synchronous requests, _Queued Writes_ are not part of this proposal. This is probably obvious, but I want to make it clear
- If the node receiving the request must forward any of the request to the Leader, then the entire request will be forwarded to the Leader, regardless of the any specified read consistency. For example, if the request contains any write-requests, and it is received by a Follower, the entire request will be forwarded to the Leader, including any read requests.

**Comment by jmordica** [comment]
That seems acceptable ^

**Comment by jvarruda** [comment]
LGTM. Thanks for the opportunity to give some feedback.

**Comment by otoolep** [comment]
_Unified Endpoint_ code is now available on master branch. Below is an example of an actual session, note the last two _curl_ requests use the `/db/request` endpoint.  Any object in the response with a key named `last_insert_id` is an "execute", otherwise a "query", assuming the `error` is not set.
```bash
~ $ curl -XPOST 'localhost:4001/db/execute?pretty&timings' -H "Content-Type: application/json" -d '[
>     "CREATE TABLE foo (id INTEGER NOT NULL PRIMARY KEY, name TEXT, age INTEGER)"
> ]'
{
    "results": [
        {
            "time": 0.000176379
        }
    ],
    "time": 0.029292095
}
~ $ curl -XPOST 'localhost:4001/db/execute?pretty&timings' -H "Content-Type: application/json" -d '[
>     "INSERT INTO foo(name, age) VALUES(\"fiona\", 20)"
> ]'
{
    "results": [
        {
            "last_insert_id": 1,
            "rows_affected": 1,
            "time": 0.000089327
        }
    ],
    "time": 0.010509629
}
~ $ curl -XPOST 'localhost:4001/db/request?pretty&timings' -H "Content-Type: application/json" -d '[
>     "INSERT INTO foo(name, age) VALUES(\"declan\", 20)",
>     "SELECT * FROM foo"
> ]'
{
    "results": [
        {
            "last_insert_id": 2,
            "rows_affected": 1,
            "time": 0.000033318
        },
        {
            "columns": [
                "id",
                "name",
                "age"
            ],
            "types": [
                "integer",
                "text",
                "integer"
            ],
            "values": [
                [
                    1,
                    "fiona",
                    20
                ],
                [
                    2,
                    "declan",
                    20
                ]
            ],
            "time": 0.000052469
        }
    ],
    "time": 0.023438583
}
~ $ curl -XPOST 'localhost:4001/db/request?pretty&timings' -H "Content-Type: application/json" -d '[
>     "SELECT COUNT(*) from baz",
>     "SELECT COUNT(*) FROM foo"
> ]'
{
    "results": [
        {
            "error": "no such table: baz"
        },
        {
            "columns": [
                "COUNT(*)"
            ],
            "types": [
                ""
            ],
            "values": [
                [
                    2
                ]
            ],
            "time": 0.000029916
        }
    ],
    "time": 0.015785677
}
~ $ 
```

**Comment by otoolep** [comment]
I believe this delivers the new feature, as specified.

**Comment by otoolep** [comment]
Available in 7.18.0.
